lzy/long-running/src/main/java/ai/lzy/longrunning/task/dao/OperationTaskDaoImpl.java

```diff
@@ -0,0 +1,319 @@
+        """.formatted(FIELDS);
+
+    //in first nested request we gather all tasks that either locked or free.
+    //in second nested request we filter result of previous request to get only free tasks
+    //and select only specific amount.
+    private static final String LOCK_PENDING_BATCH_QUERY = """
+        UPDATE operation_task
+        SET status = 'RUNNING', worker_id = ?, updated_at = now(), lease_till = now() + cast(? as interval)
+        WHERE id IN (
+            SELECT id
+            FROM operation_task
+            WHERE id IN (
+                SELECT DISTINCT ON (entity_id) id
+                FROM operation_task
+                WHERE status IN ('PENDING', 'RUNNING')
+                ORDER BY entity_id, id
+            ) AND status = 'PENDING'
+            LIMIT ?
+        )
+        RETURNING %s;
+        """.formatted(FIELDS);
+
+    private static final String RECAPTURE_OLD_TASKS_QUERY = """
+        UPDATE operation_task
+        SET updated_at = now(), lease_till = now() + cast(? as interval)
+        WHERE status = 'RUNNING' AND worker_id = ?
+        RETURNING %s;
+        """.formatted(FIELDS);
+
+    //in first nested request we select all tasks by entity_id that are either locked or free and take the first one.
+    //in second nested request we filter result of previous request to get only pending task with specific id.
+    //if we get the task then we lock it and return it.
+    private static final String TRY_LOCK_TASK_QUERY = """
+        UPDATE operation_task
+        SET status = 'RUNNING', worker_id = ?, updated_at = now(), lease_till = now() + cast(? as interval)
+        WHERE id IN (
+            SELECT id
+            FROM operation_task
+            WHERE id IN (
+                SELECT id
+                FROM operation_task
+                WHERE status IN ('PENDING', 'RUNNING') AND entity_id = ?
+                ORDER BY id
+                LIMIT 1
+            ) AND status = 'PENDING' AND id = ?
+        )
+        RETURNING %s;
+        """.formatted(FIELDS);
+
+    public static final String DELETE_QUERY = "DELETE FROM operation_task WHERE id = ?";
+    public static final String UPDATE_LEASE_QUERY = """
+        UPDATE operation_task
+        SET lease_till = now() + cast(? as interval)
+        WHERE id = ?
+        RETURNING %s
+        """.formatted(FIELDS);
+    private static final TypeReference<Map<String, Object>> MAP_TYPE_REFERENCE = new TypeReference<>() { };
+
+
+    private final Storage storage;
+    private final ObjectMapper objectMapper;
+
+    public OperationTaskDaoImpl(Storage storage, ObjectMapper objectMapper) {
+        this.storage = storage;
+        this.objectMapper = objectMapper;
+    }
+
+    @Nullable
+    @Override
+    public OperationTask get(long id, @Nullable TransactionHandle tx) throws SQLException {
+        return DbOperation.execute(tx, storage, c -> {
+            try (PreparedStatement ps = c.prepareStatement(SELECT_QUERY)) {
```

Comment:
shouldn't we add `FOR UPDATE` if `tx` is not null?

in common scenario we do the next:
```
var tx = start_tx();
var some_state = dao.get(tx);
... some business logic ...
dao.update(new_state, tx);   <-- simple UPDATE, not CAS
tx.commit();
```

if we do simple UPDATE in tx, then we should add `FOR UPDATE` to our SELECT query

Reply:
Not sure if it's necessary because we don't have long-lasting transactions that require read-and-update. It may be useful, of course, if we want to ensure that operation_task hasn't been updated by other instance (in case of parallel execution which is not desirable). So I'll revise the code and think about this problem